code/modules/crafting/craft.dm

```diff
@@ -0,0 +1,341 @@
+/datum/personal_crafting
+	var/busy
+	var/viewing_category = 1 //typical powergamer starting on the Weapons tab
+	var/list/categories = list(CAT_WEAPON,CAT_AMMO,CAT_ROBOT,CAT_FOOD,CAT_MISC,CAT_PRIMAL)
+	//var/datum/action/innate/crafting/button // Holdover from when crafting was still an action button. TODO: Remove this before I commit.
+
+
+
+
+
+/*	This is what procs do:
+	get_environment - gets a list of things accessable for crafting by user
+	get_surroundings - takes a list of things and makes a list of key-types to values-amounts of said type in the list
+	check_contents - takes a recipe and a key-type list and checks if said recipe can be done with available stuff
+	check_tools - takes recipe, a key-type list, and a user and checks if there are enough tools to do the stuff, checks bugs one level deep
+	construct_item - takes a recipe and a user, call all the checking procs, calls do_after, checks all the things again, calls del_reqs, creates result, calls CheckParts of said result with argument being list returned by deel_reqs
+	del_reqs - takes recipe and a user, loops over the recipes reqs var and tries to find everything in the list make by get_environment and delete it/add to parts list, then returns the said list
+*/
+
+
+
+
+/datum/personal_crafting/proc/check_contents(datum/crafting_recipe/R, list/contents)
+	main_loop:
```

Comment:
It's not the loop labels themselves that make it messy and less readable, it's the depth of the nested loops and nested if statements, the side effect of which is the need for the loop labels, and the depth, and therefore the need for the labels would be reduced by sub procing